src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/BufferInternal.cs

```diff
@@ -0,0 +1,271 @@
+                    *(int*)(dest + 12) = *(int*)(src + 12);
+#endif
+                    return;
+                default:
+                    break;
+            }
+
+            if (((int)dest & 3) != 0)
+            {
+                if (((int)dest & 1) != 0)
+                {
+                    *dest = *src;
+                    src++;
+                    dest++;
+                    len--;
+                    if (((int)dest & 2) == 0)
+                        goto Aligned;
+                }
+                *(short*)dest = *(short*)src;
+                src += 2;
+                dest += 2;
+                len -= 2;
+            Aligned:;
+            }
+
+#if WIN64
+                    if (((int)dest & 4) != 0)
+                    {
+                        *(int *)dest = *(int *)src;
+                        src += 4;
+                        dest += 4;
+                        len -= 4;
+                    }
+#endif
+
+#if WIN64
+                    ulong count = len / 16;
+#else
+            uint count = len / 16;
+#endif
+            while (count > 0)
+            {
+#if WIN64
+                        ((long*)dest)[0] = ((long*)src)[0];
+                        ((long*)dest)[1] = ((long*)src)[1];
+#else
+                ((int*)dest)[0] = ((int*)src)[0];
+                ((int*)dest)[1] = ((int*)src)[1];
+                ((int*)dest)[2] = ((int*)src)[2];
+                ((int*)dest)[3] = ((int*)src)[3];
+#endif
+                dest += 16;
+                src += 16;
+                count--;
+            }
+
+            if ((len & 8) != 0)
+            {
+#if WIN64
+                        ((long*)dest)[0] = ((long*)src)[0];
+#else
+                ((int*)dest)[0] = ((int*)src)[0];
+                ((int*)dest)[1] = ((int*)src)[1];
+#endif
+                dest += 8;
+                src += 8;
+            }
+            if ((len & 4) != 0)
+            {
+                ((int*)dest)[0] = ((int*)src)[0];
+                dest += 4;
+                src += 4;
+            }
+            if ((len & 2) != 0)
+            {
+                ((short*)dest)[0] = ((short*)src)[0];
+                dest += 2;
+                src += 2;
+            }
+            if ((len & 1) != 0)
+                *dest = *src;
+
+            return;
+        }
+
+        private static unsafe void SlowCopyBackwards(byte* dest, byte* src, ulong len)
+        {
+            if (len == 0) return;
+            Debug.Assert(len > 0);
+            Debug.Assert(len <= long.MaxValue);
+
+            for(long i=(long)len-1; i>=0; i--)
+            {
+                dest[i] = src[i];
+            }
+        }
+
+        private static unsafe bool AreOverlapping(byte* dest, byte* src, ulong len)
+        {
+            ulong destAddress = (ulong)dest;
+            ulong srcAddress = (ulong)src;
+            ulong srcEnd = srcAddress + len;
+            ulong destEnd = destAddress + len;
+            if (srcEnd >= destAddress && srcEnd <= destEnd)
```

Comment:
do we count for other case of the overlapping which is srcAddress lay between destAddress and destEnd but srcEnd is greater than destEnd?